source/extensions/filters/http/cdn_loop/parser.cc

```diff
@@ -0,0 +1,359 @@
+#include "extensions/filters/http/cdn_loop/parser.h"
+
+#include "common/common/statusor.h"
+
+#include "absl/status/status.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CdnLoop {
+namespace Parser {
+
+namespace {
+
+// RFC 5234 Appendix B.1 says:
+//
+// ALPHA          =  %x41-5A / %x61-7A   ; A-Z / a-z
+constexpr bool isAlpha(char c) {
+  return ('\x41' <= c && c <= '\x5a') || ('\x61' <= c && c <= '\x7a');
+}
+
+// RFC 5234 Appendix B.1 says:
+//
+// DIGIT          =  %x30-39 ; 0-9
+constexpr bool isDigit(char c) { return '\x30' <= c && c <= '\x39'; }
+
+// RFC 2234 Section 6.1 defines HEXDIG as:
+//
+// HEXDIG         =  DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
+//
+// This rule allows lower case letters too in violation of the RFC.
+constexpr bool isHexDigitCaseInsensitive(char c) {
+  return isDigit(c) || ('A' <= c && c <= 'F') || ('a' <= c && c <= 'f');
+}
+
+// RFC 7230 Section 3.2.6 defines obs-text as:
+//
+// obs-text       = %x80-FF
+constexpr bool isObsText(char c) { return 0x80 & c; }
+
+// RFC 7230 Section 3.2.6 defines qdtext as:
+//
+// qdtext         = HTAB / SP / %x21 / %x23-5B / %x5D-7E / obs-text
+constexpr bool isQdText(char c) {
+  return c == '\t' || c == ' ' || c == '\x21' || ('\x23' <= c && c <= '\x5B') ||
+         ('\x5D' <= c && c <= '\x7E') || isObsText(c);
+}
+
+// RFC 5324 Appendix B.1 says:
+//
+// VCHAR          =  %x21-7E
+//                        ; visible (printing) characters
+constexpr bool isVChar(char c) { return '\x21' <= c && c <= '\x7e'; }
+
+} // namespace
+
+ParseContext parseOptionalWhitespace(const ParseContext& input) {
```

Comment:
`skip` as a prefix is a great suggestion since this is the only "parse" function that can't fail, so it looks different than the other ones.  I went with skipOptionalWhitespace since "optional whitespace" is something the RFC mentions a bunch in contrast to other whitespace.